### PR TITLE
Removed ssh password auth

### PR DESCRIPTION
**Pull request by okankoAMZ**
# Description of the issue

This PR addresses the need to disable SSH password authentication, challenge-response authentication, and keyboard-interactive authentication on EC2 instances for enhanced security.

# Description of changes

This change adds a user_data script to the AWS EC2 instance resource in the Terraform configuration. The script performs the following actions:

1. Disables password authentication for SSH
2. Disables challenge-response authentication for SSH
3. Disables keyboard-interactive authentication for SSH
4. Restarts the SSH service to apply the changes

Note: `ChallengeResponseAuthentication` was replaced by `KbdInteractiveAuthentication` in newer versions of SSH.
Source: https://askubuntu.com/questions/1403846/ssh-challengeresponseauthentication-in-22-04-ubuntu

# License

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests

- Automated test run: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/12302119452/job/34390526816
- Manual verification on the host:

```bash
[root~]# cat /etc/ssh/sshd_config | grep Password
# Explicitly disable PasswordAuthentication. By presetting it, we
PasswordAuthentication no
PermitEmptyPasswords no
# PasswordAuthentication.  Depending on your PAM configuration,
# PAM authentication, then enable this but set PasswordAuthentication

### CODE REVIEW

File: terraform/ec2/common/linux/main.tf [hunk 1]
```diff
@@ -55,6 +55,14 @@ resource "aws_instance" "cwagent" {
   vpc_security_group_ids               = [module.basic_components.security_group]
   associate_public_ip_address          = true
   instance_initiated_shutdown_behavior = "terminate"
+  # Provide a user_data script to disable SSH password authentication
+  user_data = <<-EOT
+    #!/bin/bash
+    # Disable password authentication for SSH
+    sed -i 's/^#*PasswordAuthentication.*/PasswordAuthentication no/' /etc/ssh/sshd_config
```
Comment (the-mann) [inline terraform/ec2/common/linux/main.tf:62]: let's make sure we're adding the appropriate ChallengeResponseAuthentication as well

**Review by the-mann — CHANGES_REQUESTED** [review]

File: terraform/ec2/common/linux/main.tf (hunk 1 above)
Comment (okankoAMZ) [reply terraform/ec2/common/linux/main.tf:62]: Good call out!

File: terraform/ec2/common/linux/main.tf [hunk 2]
```diff
@@ -55,6 +55,21 @@ resource "aws_instance" "cwagent" {
...
+    
+    # Disable challenge-response authentication for SSH
+    sed -i 's/^#*ChallengeResponseAuthentication.*/ChallengeResponseAuthentication no/' /etc/ssh/sshd_config
+    
+    # Disable keyboard-interactive authentication for SSH
+    sed -i 's/^#*KbdInteractiveAuthentication.*/KbdInteractiveAuthentication no/' /etc/ssh/sshd_config
```
Comment (musa-asad) [inline terraform/ec2/common/linux/main.tf:68]: Good that you brought this in, we need to make sure to this is set to false too.

You can verify by running `sudo sshd -T` to see what the values are set to.

Comment (musa-asad) [reply terraform/ec2/common/linux/main.tf:68]: If the issue still persists, a way to overdetermine that password authentication isn't an issue is to just ensure we only allow specific ingress IPs to access the instance, which I believe is being worked on.

Moreover, there could be match/include directives that override the default values in the sshd configuration, so we might want to look out for that as well.

Comment (musa-asad) [inline terraform/ec2/common/linux/main.tf:68]:
```suggestion
    sudo sed -i 's/^#*KbdInteractiveAuthentication.*/KbdInteractiveAuthentication no/' /etc/ssh/sshd_config
```

Comment (musa-asad) [reply terraform/ec2/common/linux/main.tf:68]: We need to run with `sudo`.
```suggestion
    sudo sed -i 's/^#*KbdInteractiveAuthentication.*/KbdInteractiveAuthentication no/' /etc/ssh/sshd_config
```

**Review by musa-asad — CHANGES_REQUESTED** [review]

File: terraform/ec2/common/linux/main.tf [hunk 3]
```diff
@@ -55,6 +55,21 @@ resource "aws_instance" "cwagent" {
...
+
+    # Restart SSH service to apply changes
+    systemctl restart sshd
+  EOT
```
Comment (musa-asad) [inline terraform/ec2/common/linux/main.tf:72]: We need to run these commands with `sudo`.
```suggestion
  user_data = <<-EOT
    #!/bin/bash
    # Disable password authentication for SSH
    sudo sed -i 's/^#*PasswordAuthentication.*/PasswordAuthentication no/' /etc/ssh/sshd_config
    
    # Disable challenge-response authentication for SSH
    sudo sed -i 's/^#*ChallengeResponseAuthentication.*/ChallengeResponseAuthentication no/' /etc/ssh/sshd_config
    
    # Disable keyboard-interactive authentication for SSH
    sudo sed -i 's/^#*KbdInteractiveAuthentication.*/KbdInteractiveAuthentication no/' /etc/ssh/sshd_config

    # Restart SSH service to apply changes
    sudo systemctl restart sshd
  EOT
```

Comment (musa-asad) [inline terraform/ec2/common/linux/main.tf:72]: We need to run these commands with `sudo`.
```suggestion
  user_data = <<-EOT
    #!/bin/bash
    # Disable password authentication for SSH
    sudo sed -i 's/^#*PasswordAuthentication.*/PasswordAuthentication no/' /etc/ssh/sshd_config
    
    # Disable challenge-response authentication for SSH
    sudo sed -i 's/^#*ChallengeResponseAuthentication.*/ChallengeResponseAuthentication no/' /etc/ssh/sshd_config
    
    # Disable keyboard-interactive authentication for SSH
    sudo sed -i 's/^#*KbdInteractiveAuthentication.*/KbdInteractiveAuthentication no/' /etc/ssh/sshd_config

    # Restart SSH service to apply changes
    sudo systemctl restart sshd
  EOT
```

**Review by musa-asad — CHANGES_REQUESTED** [review]

**Review by musa-asad — APPROVED** [review]

**Pull request merged**